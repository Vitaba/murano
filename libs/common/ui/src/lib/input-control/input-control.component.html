<div [class]="styles.container">
    <input [class]="styles.input" [type]="config.type">
  </div>