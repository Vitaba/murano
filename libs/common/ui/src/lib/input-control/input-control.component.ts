// tslint:disable: no-unsafe-any no-any
import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnChanges, SimpleChanges } from '@angular/core';
import { FormControl } from '@angular/forms';
import { GuardError, InputControlConfig, InputControlStyles, isInputControlConfig, isInputControlStyles } from '@vitaba/common-utils';
import { BehaviorSubject } from 'rxjs';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-input-control',
  styleUrls: ['./input-control.component.scss'],
  templateUrl: './input-control.component.html',
})
export class InputControlComponent implements OnChanges, AfterViewInit {
  public changes!: BehaviorSubject<SimpleChanges>;
  public errors!: GuardError;

  @Input() public config: InputControlConfig = {
    fControl: new FormControl('sample'),
    placeholder: 'placeholder',
    type: 'text',
  };

  @Input() public styles: InputControlStyles = {
    container: `md:w-2/3`,
    // tslint:disable-next-line:ter-max-len
    input: `bg-gray-200 appearance-none border-2 border-gray-200 rounded w-auto py-2 px-4 text-gray-700 leading-tight focus:outline-none focus:bg-white focus:border-purple-500`,
  };

  public constructor(
    private readonly _changeDetectorRef: ChangeDetectorRef,
  ) {}

  public ngOnChanges(changes: SimpleChanges): void {
    if (!this.changes) {
      this.changes = new BehaviorSubject(changes);
    }
    this.changes.next(changes);
  }

  public ngAfterViewInit(): void {
    if (!this.changes) {
      this.changes = new BehaviorSubject(null);
    }
    this.changes.subscribe((changes: SimpleChanges) => {
      if (changes.config) {
        const validation =
        this.validateInputControlConfig(
          changes.config.currentValue as InputControlConfig);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }

      if (changes.styles) {
        const validation =
        this.validateInputControlStyles(
          changes.styles.currentValue as InputControlStyles);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }
      this._changeDetectorRef.detectChanges();
    });
  }

  public validateInputControlConfig(value: InputControlConfig) {
    const validations = [{
      name: 'placeholder',
      type: 'string',
    },
      {
        name: 'type',
        type: 'string',
      },
    ];

    if (value.type === 'number' || value.type === 'range') {
      validations.push({
        name: 'min',
        type: 'number',
      });
      validations.push({
        name: 'max',
        type: 'number',
      });
      validations.push({
        name: 'step',
        type: 'number',
      });
    }

    const validation = isInputControlConfig(value, validations);

    return !validation.valid ? validation : undefined;
  }

  public validateInputControlStyles(value: InputControlStyles) {
    const validations = [
      {
        name: 'container',
        type: 'string',
      },
      {
        name: 'input',
        type: 'string',
      },
    ];

    const validation = isInputControlStyles(value, validations);

    return !validation.valid ? validation : undefined;
  }
}
