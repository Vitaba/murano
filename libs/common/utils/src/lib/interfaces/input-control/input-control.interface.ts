import { FormControl } from '@angular/forms';

export interface InputControlConfig {
  placeholder: string;
  type: string;
  fControl: FormControl;
  disabled?: boolean;
  mask?: string;
  maskConfig?: {
    thousandSeparator?: string;
    separatorLimit?: string;
  };
  min?: number;
  max?: number;
  step?: number;
}

export interface InputControlStyles {
  container: string;
  input: string;
}
