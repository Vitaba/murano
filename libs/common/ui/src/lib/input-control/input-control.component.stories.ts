// tslint:disable:no-default-export object-literal-sort-keys ter-max-len

import { FormControl } from '@angular/forms';
import { text } from '@storybook/addon-knobs';
import { InputControlComponent } from './input-control.component';

export default {
  title: 'Input Control',
};

const config = {
  fControl: new FormControl('sample'),
  placeholder: 'placeholder',
  type: 'text',
};

const styles = {
  container: `md:w-2/3`,
  input: `bg-gray-200 appearance-none border-2 border-gray-200 rounded w-auto py-2 px-4 text-gray-700 leading-tight focus:outline-none focus:bg-white focus:border-purple-500`,
};

export const basic = () => ({
  moduleMetadata: {
    declarations: [InputControlComponent],
    imports: [],
  },
  props: {
    config_placeholder: text('placeholder', config.placeholder, 'config'),
    config_type: text('type', config.type, 'config'),
    styles_container: text('container', styles.container, 'styles'),
    styles_input: text('input', styles.input, 'styles'),
  },
  template: `
  <vitaba-input-control
    [config]="{
    placeholder: config_placeholder,
    type: config_type
    }"
    [styles]="{
    container: styles_container,
    input: styles_input
    }"
    > </vitaba-input-control>
  `,
});
