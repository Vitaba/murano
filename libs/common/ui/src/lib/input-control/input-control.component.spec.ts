
// tslint:disable: no-any component-max-inline-declarations ter-max-len
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { NgxMaskModule } from 'ngx-mask';
import { InputControlComponent } from './input-control.component';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-input-control-sample-implementation',
  template: `
    <vitaba-input-control [config]="config" [styles]="styles">
    </vitaba-input-control>
  `,
})
class TestHostComponent {
  public styles = {
    container: `md:w-2/3`,
    input: `bg-gray-200 appearance-none border-2
    border-gray-200 rounded w-auto py-2 px-4
    text-gray-700 leading-tight focus:outline-none
    focus:bg-white focus:border-purple-500`,
  };
  public data = {
    description: 'Useful resources and assets for users.',
    title: 'Lorem Ipsum',
  };
  public config = {
    fControl: new FormControl('sample'),
    placeholder: 'placeholder',
    type: 'text',
  };
}

describe('InputControlComponent', () => {
  let component: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestHostComponent, InputControlComponent],
      imports: [ReactiveFormsModule, NgxMaskModule.forRoot()],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should be default', () => {
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should throw errors when config is not in right format', () => {
    component.config = {
      fControl: new FormControl('name'),
      placeholder: 1,
      type: 1,
    } as any;
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
