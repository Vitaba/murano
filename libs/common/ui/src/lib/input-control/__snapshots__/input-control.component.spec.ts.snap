// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputControlComponent should be default 1`] = `
<vitaba-input-control-sample-implementation
  config={[Function Object]}
  data={[Function Object]}
  styles={[Function Object]}
>
  <vitaba-input-control>
    
    
    <div
      class="md:w-2/3"
    >
      <div
        class="relative"
      />
      
      <input
        class="bg-gray-200 appearance-none border-2 border-gray-200 rounded w-auto py-2 px-4 text-gray-700 leading-tight focus:outline-none focus:bg-white focus:border-purple-500 ng-untouched ng-pristine ng-valid"
        placeholder="placeholder"
        type="text"
      />
      
    </div>
    
  </vitaba-input-control>
</vitaba-input-control-sample-implementation>
`;

exports[`InputControlComponent should throw errors when config is not in right format 1`] = `
<vitaba-input-control-sample-implementation
  config={[Function Object]}
  data={[Function Object]}
  styles={[Function Object]}
>
  <vitaba-input-control>
    
     {
    "valid": false,
    "guard": "isInputConfig",
    "value": {
      "fControl": {
        "validator": null,
        "asyncValidator": null,
        "pristine": true,
        "touched": false,
        "_onDisabledChange": [
          null
        ],
        "_onChange": [
          null
        ],
        "_pendingValue": "name",
        "value": "name",
        "status": "VALID",
        "errors": null,
        "valueChanges": {
          "_isScalar": false,
          "observers": [],
          "closed": false,
          "isStopped": false,
          "hasError": false,
          "thrownError": null,
          "__isAsync": false
        },
        "statusChanges": {
          "_isScalar": false,
          "observers": [],
          "closed": false,
          "isStopped": false,
          "hasError": false,
          "thrownError": null,
          "__isAsync": false
        }
      },
      "placeholder": 1,
      "type": 1
    },
    "message": "Is not a input config",
    "errors": [
      {
        "valid": false,
        "guard": "hasProperties",
        "value": {
          "fControl": {
            "validator": null,
            "asyncValidator": null,
            "pristine": true,
            "touched": false,
            "_onDisabledChange": [
              null
            ],
            "_onChange": [
              null
            ],
            "_pendingValue": "name",
            "value": "name",
            "status": "VALID",
            "errors": null,
            "valueChanges": {
              "_isScalar": false,
              "observers": [],
              "closed": false,
              "isStopped": false,
              "hasError": false,
              "thrownError": null,
              "__isAsync": false
            },
            "statusChanges": {
              "_isScalar": false,
              "observers": [],
              "closed": false,
              "isStopped": false,
              "hasError": false,
              "thrownError": null,
              "__isAsync": false
            }
          },
          "placeholder": 1,
          "type": 1
        },
        "errors": [
          {
            "value": {
              "name": "placeholder",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          },
          {
            "value": {
              "name": "type",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          }
        ]
      }
    ]
  } 
    
  </vitaba-input-control>
</vitaba-input-control-sample-implementation>
`;
